rfcs/theme-shape.md

```diff
@@ -0,0 +1,228 @@
+
+### Master
+
+Current implementation in `@fluentui/react-theme@9.0.0-alpha.22`.
+
+- 1200 CSS variables in a theme = ~650 global tokens + ~550 alias tokens
+- alias token implemented as a "reference" to a global token: `--alias-token: var(--global-token)`.
+- **Each reflow takes ~150ms (1500 ms total for the 10 mounts)**
+
+![Flamechart - Master](theme-shape/perf-master.png)
+
+### Inline values to alias tokens.
+
+- Still the same 1200 CSS variables (650 global + 550 alias)
+- Instead of "referencing" global tokens in the alias tokens, inline the values: `--alias-token: #fff`.
+- **Each reflow takes ~90ms (1000 ms total for the 10 mounts)**
+
+![Flamechart - Inline values to alias tokens](theme-shape/perf-inline-alias.png)
+
+### Reduce number of tokens (CSS variables)
+
+- Removed global color tokens, number of CSS variables reduced to 550.
+- **Each reflow takes ~50ms (600 ms total for the 10 mounts)**
+
+![Flamechart - Reduce number of tokens](theme-shape/perf-no-global.png)
+
+### Remove even more tokens
+
+It seems that the rendering performance depends on the number of CSS variables in the DOM. This is an experiment to remove even more variables to verify the hypothesis.
+
+- Removed alias shared color tokens, number of CSS variables reduced to 160.
+- **Each reflow takes ~20ms (350 ms total for the 10 mounts)**
+
+![Flamechart - Remove even more tokens](theme-shape/perf-no-colors.png)
+
+### Hash CSS variables
+
+In inline aliases scenario, there are 1200 CSS variables with names like `--alias-color-neutral-neutralStrokeAccessibleHover`. The total string length of CSS variable names (including the leading `--`) is 39,252 characters - is that affecting the perf as well? We used 7-char hashes for variable names to check for any difference - this reduced the string length to 10,354 characters. There is no difference in the performance.
+
+- **Each reflow takes ~90ms (1000 ms total for the 10 mounts)**
+
+**The perf example intentionally uses precomputed hashes to keep the hash computation out of the measured code path.**
+
+![Flamechart - Hash CSS variables](theme-shape/perf-hash-variables.png)
+
+### Results Summary
+
+We also used flamegrill to compare the approaches. Inlined aliases is used as a baseline all other approaches are compared to. For each experiment we run the test 10 times for both the experiment and the baseline to get comparable numbers.
+
+![Flamegrill - Results Summary](theme-shape/perf-flamegrill.png)
+
+## Proposed changes to the theme shape
+
+1. [Inline values in alias tokens](#inline-alias)
+2. [Do not inject global color tokens as CSS variables](#no-global-tokens)
+3. [Merge global and alias tokens to a single object](#merge-global-alias)
+4. [Flatten the tokens object](#flatten-tokens)
+5. [Hash variable names](#hash-variables)
+6. [Be cautious when adding new tokens to the theme](#be-cautious)
+7. [Get rid of all CSS variables](#no-css-vars)
```

Comment:
Would getting rid of all CSS variables be a problem if a dev has a static stylesheet for globals in their app and would like to use some of the theme variables?

Reply:
CSS variables representing the design tokens should never be used directly. That would make the CSS variable names part of the API surface, we want to keep them encapsulated.

The way how we plan to support SASS/LESS is the following:
1. We currently use a token pipeline to transform design tokens to a Typescript object which is used inside the library and injected as CSS variables to DOM by a `FluentProvider` - these names are currently predictable, we might change that to use hashed variable names to enforce the encapsulation (after beta).
2. The same pipeline will be used to generate a list of SASS variables which will map to `FluentProvider` CSS variables:
```scss
$neutralbackground2-fill-color-hover: var(--neutralbackground2-fill-color-hover)
```
3. The SASS styles will reference that SASS variables.
4. There are two possible ways how the CSS variables will be injected into the DOM:
    - either you are inside the DOM subtree managed by the `FluentProvider` which is then responsible for the variables
    - or you use a static CSS generated by the pipeline and it will be you responsibility to inject it + swap it to switch themes

Without CSS variables, the process would be pretty much the same, with the following differences:
- in step 2, instead of pointing to the CSS variable, the SASS variable would point to the direct value
- in step 4, you would also need to swap the stylesheet with your static styles to match the required theme.

But good catch, even though we do not plan to get rid of the CSS variables with now, I will add the static stylesheets problem to the list of possible complications.